### PR TITLE
Modify ActiveRecord eql? to Handle STI classes

### PR DESCRIPTION
**Pull request by CyborgMaster**
This is the implementation of my idea from issue #23760.  See the issue for a full discussion of the bug and motivation for the change.

### CODE REVIEW

**Comment by maclover7** [comment]
Can you add some tests to show how the behavior has changed?

**Comment by CyborgMaster** [comment]
I'd be happy to.  I'm not really sure how to write tests for ActiveRecord, it would seem like I would need to create temporary tables and classes to make it work.  I've never contributed to Rails before.  Any suggestions on where I should look for examples on how to get started?

**Comment by maclover7** [comment]
Based on that you are changing the `core.rb` file in Active Record, you would want to add your tests to `activerecord/test/cases/core_test.rb`. There are already a lot of models defined `activerecord/test/models` -- try and find one that uses STI, and then from there you can right your `==` test(s). :)

**Comment by CyborgMaster** [comment]
Thanks for helping me get started.  I'm trying now to get the test file to run so I can get a good red-green cycle.  I'll let you know if I have any more questions.

**Comment by CyborgMaster** [comment]
I can't seem to get any of the tests to run.  Is there some documentation anywhere describing how get things test up to run tests?  I don't really need to run the fully test suite, just that one file.

**Comment by CyborgMaster** [comment]
Ah hah!  I think I've got it.  It took me a bit.  The instructions on the RailsGuide left a few things out about dependencies, but I'm pretty sure I'm up an running now.

**Comment by CyborgMaster** [comment]
Ok.  I think I've got a test that demonstrates the new behavior.  Would you please take a look at it and see if it meets your requirements.  Now that I have got the test environment set up, it will be easy to modify if you want something added / changed.

**Comment by maclover7** [comment]
This looks ok to me. Now you just have to wait for someone that's higher than me on the totem pole to review your PR. :)

**Comment by CyborgMaster** [comment]
Awesome.  Thanks a ton for your help.

**Comment by mdesantis** [comment]
Hello, is there any chance to merge this into Rails 5? It's a welcome patch to me :)

**Comment by nlcurry** [comment]
Any progress on this issue?

**Comment by CyborgMaster** [comment]
I'd love to get it merged too.  I'm still using a monkey patch in production.  Would be great to get rid of it.

**Comment by phaedryx** [comment]
I would love to get this merged.

File: activerecord/lib/active_record/core.rb [hunk 1]
```diff
@@ -413,7 +413,7 @@ def encode_with(coder)
     # models are still comparable.
     def ==(comparison_object)
       super ||
-        comparison_object.instance_of?(self.class) &&
+        comparison_object.is_a?(self.class.base_class) &&
```
Comment (gmcgibbon) [inline activerecord/lib/active_record/core.rb]: Do we want to support `STISuperClass.first == STISubClass.first`? You would need to do something like `comparison_object.class.base_class == self.class.base_class` to match equality in both combinations.

**Review by gmcgibbon — COMMENTED** [review]
Still an issue in Rails 6 (e925cb4d856088a815bf4a0cf27518d01bb4029d). I think this makes sense. 👍 One comment about equality order though.

File: activerecord/lib/active_record/core.rb (hunk 1 above)
Comment (CyborgMaster) [reply activerecord/lib/active_record/core.rb]: I don't think we need to move to `comparison_object.class.base_class == self.class.base_class` in order to work for both comparisons.  The comparison object in both cases is going to be a subclass of the top level STI base class.  So the `is_a?` operation should always work.  In your example `self.class.base_class` will be `STISuperClass` and `STISuperClass.first.is_a? STISuperClass` and `STISubClass.first.is_a? STISuperClass` will both be true.

Comment (gmcgibbon) [reply activerecord/lib/active_record/core.rb]: Ah! You are correct. I thought `#is_a?` and `#instance_of?` were synonymous, but they are not. 👍 In that case please rebase and squash your commits for this to be mergeable. Also I think this is CHANGELOG worthy.

**Comment by CyborgMaster** [comment]
@gmcgibbon, Rebased, squashed, and CHANGELOG entry added!  Anything else I should do (this is my first contribution to Rails, so I may have missed something)?

**Comment by matthewd** [comment]
I'm really not sure it makes sense to consider such objects `==` equal... the only thing we know about them is that the developer has given them different classes: if they were truly equivalent, why change the class at all?

I'm much more sure that they should not be `eql?` equal.

**Comment by mdesantis** [comment]
I think at STI records as equal, since they refer to the same table. They can implement different methods, but in this case there is an exception IMHO. I'm expressing a point of view based on my experience: every time in the apps where I worked I added STI it was logical to have object equality.

**Comment by CyborgMaster** [comment]
I agree that treating objects of different classes as equals is unusual, but I believe it necessary for STI.  The ActiveRecord definition of `==` treats objects as equal if they refer to the same database record.  Ex: 

``` ruby
c1 = Client.first # #<Client:0x007f99ef748818 ... >
c2 = Client.first  # #<Client:0x007f99ef748818 ... >
                            # ^- note different object id
c1 == c2 # true                                                                
```

This rule should hold true for STI classes as well. 

When STI subclasses are not equal it also causes a weird corner case when combining changing the type of a record and reassigning it to a association:

Quoted from #23760

> ``` ruby
> class CommunicationSetting < ActiveRecord::Base
>   has_many :filters
> end
> 
> class Filter < ActiveRecord::Base
>   belongs_to :communication_setting
> end
> 
> class AppointmentResourceFilter < Filter
> end
> 
> class AppointmentTypeFilter < Filter
> end
> 
> c = CommunicationSetting.first
> c.filters # => [#<AppointmentResourceFilter:0x007f9ac80d2c70 id: 1, type: "AppointmentResourceFilter">]
> 
> f1 = c.filters.first
> f1 # => #<AppointmentResourceFilter:0x007f9ac80d2c70 id: 1, type: "AppointmentResourceFilter">
> 
> f2 = f1.becomes!(AppointmentTypeFilter)
> f2 # => #<AppointmentTypeFilter:0x007f9ac804e560 id: 1, type: "AppointmentTypeFilter">
> 
> f1 == f2 # => false
> 
> c.filters = [f2]
> ```
> 
> Normally if you replace the association on an active record object it realizes if records are the same (based on having the same primary key) and then executes an update query.  If there are removed records, it deletes them and if there are new ones, it inserts them.  
> 
> However in this case it is very, very confused.  Because the `==` operator returns false, ActiveRecord thinks that the "old" record needs to be deleted.  Then it tries to save the new record, which reports that is has already been persisted because it was loaded from the database, so ActiveRecord tries to perform an update query, which doesn't work:
> 
> ``` sql
> DELETE FROM `filters` WHERE `filters`.`id` = 1
> UPDATE `filters` SET `type` = 'AppointmentTypeFilter', `updated_at` = '2016-02-18 17:20:07' WHERE `filters`.`id` = 1
> ```
> 
> As you can see from the above, it deletes record 1 and then tries to update record 1, which now doesn't exist.  
>

**Comment by CyborgMaster** [comment]
I just rebased and fixed conflicts

**Comment by yskkin** [comment]
At least, `is_a?` is not enough to check two classes are in STI relation:

```ruby
class Article < ApplicationRecord
end

class ArchivedArticle < Article
  self.table_name = 'archived_articles'
end
```

I didn't test but gems like https://github.com/mvdamme/dbview_cti may break...

**Comment by gmcgibbon** [comment]
🤔 There have been some good points made for why this change shouldn't be merged. Should we settle for using [`#becomes`](https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-becomes) before equality checks or try comparing `record.class.table_name`s instead of using `#instance_of?` / `#is_a?`?

**Comment by CyborgMaster** [comment]
Any thought on my above response the concerns?

**Comment by CyborgMaster** [comment]
I'm still interested in getting this merged.  Are there any thoughts on my last response?

**Comment by phaedryx** [comment]
I'm still interested in this too

**Comment by rafaelfranca** [comment]
The documentation is very clear about `==` not returning true to objects with different types:

```
# Returns true if +comparison_object+ is the same exact object, or +comparison_object+
# is of the same type and +self+ has an ID and it is equal to +comparison_object.id+.
```

So I agree with @matthewd, this change should not be merged.

I don't think the #23760 should be fixed by changing the meaning of `==`, maybe the association method should use a different method for database identity instead of object identity.

**Pull request closed without merging**